pkg/agent/controller/networkpolicy/l7engine/reconciler_test.go

```diff
@@ -0,0 +1,227 @@
+// Copyright 2022 Antrea Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package l7engine
+
+import (
+	"strings"
+	"testing"
+
+	"github.com/spf13/afero"
+	"github.com/stretchr/testify/assert"
+
+	v1beta "antrea.io/antrea/pkg/apis/controlplane/v1beta2"
+)
+
+var (
+	startSuricataFnCalled bool
+
+	suricataScFnCalledWithReloadTenant            bool
+	suricataScFnCalledWithRegisterTenantHandler   bool
+	suricataScFnCalledWithRegisterTenant          bool
+	suricataScFnCalledWithUnregisterTenantHandler bool
+	suricataScFnCalledWithUnregisterTenant        bool
+)
+
+func TestConvertProtocolHTTP(t *testing.T) {
+	testCases := []struct {
+		name     string
+		http     *v1beta.HTTPProtocol
+		expected string
+	}{
+		{
+			name:     "without host,method,path",
+			http:     &v1beta.HTTPProtocol{},
+			expected: "",
+		},
+		{
+			name: "with host,method,path",
+			http: &v1beta.HTTPProtocol{
+				Host:   "www.google.com",
+				Method: "GET",
+				Path:   "index.html",
+			},
+			expected: "http.uri; content:\"index.html\"; http.method; content:\"GET\"; http.host; content:\"www.google.com\";",
+		},
+		{
+			name: "with host",
+			http: &v1beta.HTTPProtocol{
+				Host: "www.google.com",
+			},
+			expected: "http.host; content:\"www.google.com\";",
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			assert.Equal(t, tc.expected, convertProtocolHTTP(tc.http))
+		})
+	}
+}
+
+func newFakeReconciler() *Reconciler {
+	defaultFS.Create(defaultSuricataConfigPath)
+
+	fe := NewReconciler()
+	fe.suricataScFn = func(scCmd string) (*scCmdRet, error) {
+		if strings.HasPrefix(scCmd, "reload-tenant") {
+			suricataScFnCalledWithReloadTenant = true
+		} else if strings.HasPrefix(scCmd, "register-tenant-handler") {
+			suricataScFnCalledWithRegisterTenantHandler = true
+		} else if strings.HasPrefix(scCmd, "register-tenant") {
+			suricataScFnCalledWithRegisterTenant = true
+		} else if strings.HasPrefix(scCmd, "unregister-tenant-handler") {
+			suricataScFnCalledWithUnregisterTenantHandler = true
+		} else if strings.HasPrefix(scCmd, "unregister-tenant") {
+			suricataScFnCalledWithUnregisterTenant = true
```

Comment:
It looks a little inefficient to write and maintain this mock functions and the global variables.
Consider this:
```
type fakeSuricata struct {
	calledScCommands      sets.String
	startSuricataFnCalled bool
}

func newFakeSuricata() *fakeSuricata {
	return &fakeSuricata{
		calledScCommands:      sets.NewString(),
		startSuricataFnCalled: false,
	}
}

func (f *fakeSuricata) suricataScfunc(scCmd string) (*scCmdRet, error) {
	f.calledScCommands.Insert(scCmd)
	return &scCmdRet{Return: scCmdOK}, nil
}

func (f *fakeSuricata) startSuricataFn() {
	f.startSuricataFnCalled = true
	defaultFS.Create(suricataCommandSocket)
}

fe := NewReconciler()
fs := newFakeSuricata()
fe.suricataScFn = fs.suricataScfunc
fe.startSuricataFn = fs.startSuricataFn

...
expectedScCommands := sets.NewString("register-tenant", "register-tenant-handler", ...)
assert.Equal(t, expectedScCommands, fs.calledScCommands)
assert.True(t, fs.startSuricataFnCalled)

```